src/main/java/org/assertj/core/api/AbstractAssert.java

```diff
@@ -830,4 +834,23 @@ private void propagateAssertionInfoFrom(AbstractAssert<?, ?> assertInstance) {
     return usingRecursiveComparison(new RecursiveComparisonConfiguration());
   }
 
+  @SuppressWarnings("unchecked")
+  protected <ASSERT extends AbstractAssert<?, ?>> ASSERT extracting(String propertyOrField,
+                                                                    AssertFactory<Object, ASSERT> assertFactory) {
```

Comment:
> > I had a bit of a play with this - I think it's a good idea, but to do it requires that the method signature of extracting() be changed to parameterize the type of the extracted property, otherwise the compiler won't allow the cast in (ASSERT) assertFactory.createAssert() to convert to the appropriate return type.
> 
> Your example compiles, but this one looks risky to me:
> 
> ```java
> T value = (T)byName(propertyOrField).apply(actual);
> ```
> 
> How can you be sure that you are casting to the correct type? `Extractors.byName()` has no knowledge about the actual field/property type at compile time, that's why it should always give back an `Object`. Am I missing something?

Correct, there is the potential for a `ClassCastException` here at run time if you're not careful. I'm not too concerned about this, for two reasons:

1. We're testing - we are always expecting exceptions to be thrown and we're ready to catch them. It won't cause errors in production code, and I actually think it's useful if an error in the test code manifests as some other kind of exception (like `ClassCastException`) rather than `AssertionError`.
2. I'm pretty sure this problem existed in the original code anyway, just in a different place - if you pass the wrong type of `AssertFactory` that is expecting an `ACTUAL` of a particular type, then you'll get a class cast exception. I guess that's why the signature currently specifies an Object for the "ACTUAL" type of the assertion factory, so that this can't happen, but this means that the assertThat assertion factories cannot be used as-is because their constructors all expect a particular type and not a plain `Object`. So I'm guessing in practice people would use one of the existing `assertThat()s` and do an unchecked cast on the result to `AssertFactory<Object, ASSERT>` instead, which has the same potential to cause for a `ClassCastException`. Or else instead of a cast you wrap it in a lambda expression and cast the value of `actual` to the narrowed type required by `assertThat()`, but again this has the same potential for a `ClassCastException`.

At least, that's how I see it - let me know if I'm missing something.

Reply:
> That's a challenge I tried to solve sometime ago (#1517) but I haven't found any clean solution without heavily changing `assertThat` entry points 🙁 ...
> At that time, together with @joel-costigliola we decided to go with the pragmatic wildcard type parameter and `(ASSERT)` cast.

Hmm, fair enough - glad I'm not the only one who's struggled with this. :smile: My gut feeling is still that this ''should'' be possible - either we're missing something, or else there is some deficiency in Java that prevents this from happening.

Reply:
It is indeed possible, but requires some changes against the current pattern in AssertJ. Maybe it's a change we could consider for the next major release.

Reply:
> Correct, there is the potential for a `ClassCastException` here at run time if you're not careful. I'm not too concerned about this, for two reasons:
> 
>     1. We're testing - we are always expecting exceptions to be thrown and we're ready to catch them. It won't cause errors in production code, and I actually think it's useful if an error in the test code manifests as some other kind of exception (like `ClassCastException`) rather than `AssertionError`.

I'm not fully convinced about this. It's true that we are under testing scope, but the way I see it is that AssertJ should give back `AssertionError`s as much as it can from standard flows and other exceptions in case of misusage (i.e., null parameters). Having an api which allows parameterized types but can give back `ClassCastException` if you provide the wrong type is too much from my point of view, the user would need to understand what happens behind that api to use it properly and I don't see it as ideal.

>     2. I'm pretty sure this problem existed in the original code anyway, just in a different place - if you pass the wrong type of `AssertFactory` that is expecting an `ACTUAL` of a particular type, then you'll get a class cast exception.

`AssertFactory` per se does not guarantee any type safety and does not impose any constraint for the implementations in the javadoc. Everything depends on how the implementation declares the interface with the corresponding type parameters and the compiler will do the rest of the job.

> I guess that's why the signature currently specifies an Object for the "ACTUAL" type of the assertion factory, so that this can't happen, but this means that the assertThat assertion factories cannot be used as-is because their constructors all expect a particular type and not a plain `Object`.

I try to explain it, but once I add tests to this PR they will probably be a better description 🙂 

`Assertions.assertThat()` methods cannot be used (except the `ObjectAssert` one) because there is no way the protected `extracting()` can give you more than `Object` at compile time.

If a different type than `ObjectAssert` is still needed at compile type, `extracting` can be called directly with an `InstanceOfAssertFactory` (either one declared in `InstanceOfAssertFactories` or a new `InstanceOfAssertFactory` instance):

```java
super.extracting(propertyOrField, InstanceOfAssertFactories.DOUBLE);
```

If actual doesn't match the factory expected type, a `ClassCastException` will be thrown inside `InstanceOfAssertFactory`, which I think is a clearer place rather than in the protected `extracting`.

Of course the caller is also free to provide a custom implementation of `AssertFactory`, but has to fully take care of the wanted behavior.

> So I'm guessing in practice people would use one of the existing `assertThat()s` and do an unchecked cast on the result to `AssertFactory<Object, ASSERT>` instead, which has the same potential to cause for a `ClassCastException`. Or else instead of a cast you wrap it in a lambda expression and cast the value of `actual` to the narrowed type required by `assertThat()`, but again this has the same potential for a `ClassCastException`.

As mentioned above, if the target is to switch type an `InstanceOfAssertFactory` should be provided, either one from `InstanceOfAssertFactories` or a custom one. This can give you a `ClassCastException` but it should be for internal usage, an end user should receive an `isInstanceOf` `AssertionError` if he is the target for it. See the `asInstanceOf` implementation, it asserts the type first and does the conversion afterwards.

https://github.com/joel-costigliola/assertj-core/blob/476fe5540c20900928fa045522bd9f35f8c33ed5/src/main/java/org/assertj/core/api/AbstractAssert.java#L356-L360

Reply:
> > 1. We're testing - we are always expecting exceptions to be thrown and we're ready to catch them. It won't cause errors in production code, and I actually think it's useful if an error in the test code manifests as some other kind of exception (like `ClassCastException`) rather than `AssertionError`.
> 
> I'm not fully convinced about this. It's true that we are under testing scope, but the way I see it is that AssertJ should give back `AssertionError`s as much as it can from standard flows and other exceptions in case of misusage (i.e., null parameters). Having an api which allows parameterized types but can give back `ClassCastException` if you provide the wrong type is too much from my point of view, the user would need to understand what happens behind that api to use it properly and I don't see it as ideal.
>

I guess we can agree to disagree on this one, or at least agree to defer it to another issue for discussion so as not to clutter this one. :smile:

> > 1. I'm pretty sure this problem existed in the original code anyway, just in a different place - if you pass the wrong type of `AssertFactory` that is expecting an `ACTUAL` of a particular type, then you'll get a class cast exception.
> 
> `AssertFactory` per se does not guarantee any type safety and does not impose any constraint for the implementations in the javadoc. Everything depends on how the implementation declares the interface with the corresponding type parameters and the compiler will do the rest of the job.

Yes, it depends on the `AssertFactory`'s implementation, but my point is that at some point *any* implementation will either need to be so generic as to only work with `Object`s, or else it would need to perform a narrowing cast to a particular type, and that narrowing cast could throw `ClassCastException`.

> > I guess that's why the signature currently specifies an Object for the "ACTUAL" type of the assertion factory, so that this can't happen, but this means that the assertThat assertion factories cannot be used as-is because their constructors all expect a particular type and not a plain `Object`.
> 
> I try to explain it, but once I add tests to this PR they will probably be a better explanation 🙂
> 
> `Assertions.assertThat()` methods cannot be used (except the `ObjectAssert` one) because there is no way the protected `extracting()` can give you more than `Object` at compile time.
> 
> If a different type than `ObjectAssert` is still needed at compile type, `extracting` can be called directly with an `InstanceOfAssertFactory` (either one declared in `InstanceOfAssertFactories` or a new `InstanceOfAssertFactory` instance):
> 
> ```java
> super.extracting(propertyOrField, InstanceOfAssertFactories.DOUBLE);
> ```
> 
> If actual doesn't match the factory expected type, a `ClassCastException` will be thrown inside `InstanceOfAssertFactory`, which I think is a clearer place rather than in the protected `extracting`.
>

Actually, even in the way I proposed it, the `ClassCastException` will be thrown in `InstanceOfAssertFactory` at the same place that it is current (and not in `extracting()`).

That is because the actual type of `T` is not available at runtime - the compiler replaces `T` with the erasure of that type, which means (because `T` is unbounded) it means that it compiles to: `Object value = byName(propertyOrField).apply(actual);` - ie, the same as it was previously. This line can obviously never throw a `ClassCastException` (that is why the compiler flags it with an "unchecked cast" warning, because the compiler has no way to check the type it is casting to).

In other words, at runtime it would be the same as the current situation. The only thing that this cast does is allow you to have a bit better compile-time type safety on the `AssertFactory` instance.